vector.hpp

```diff
@@ -0,0 +1,276 @@
+ *    You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <array>
+
+#ifndef VECTOR_HPP_
+#define VECTOR_HPP_
+
+template <typename T, int dimension>
+class Vector {
+ public:
+  static_assert(dimension>0, "Dimension cannot be non-zero");
+  
+  /**
+   * @brief    Constructors for the class for a zero vector.
+   */
+  Vector();
+
+  /**
+   * @brief    Constructors for the class for a particular vector.
+   */
+  Vector(const T vector[]);
+
+  /**
+   * @brief    Conversion from a vector type to another.
+   */
+  template <typename U>
+  Vector(const Vector<U,dimension>& rhs);
+
+  /**
+   * @brief    Conversion from a vector type by assignment.
+   */
+  template <typename U>
+  Vector<T,dimension>& operator=(const Vector<U,dimension>& rhs);
+
+  /**
+   * @brief    For assigning values to entries in a vector.
+   */
+  T& operator[] (int index);
+
+  /**
+   * @brief    For accessing entries in the vector.
+   */
+  T operator[] (int index) const;
+
+  Vector<T,dimension> operator-() const;
+  Vector<T,dimension>& operator+=(const Vector<T,dimension>& rhs);
+  Vector<T,dimension>& operator-=(const Vector<T,dimension>& rhs);
+
+  /**
+   * @brief    Addition or subtraction of every
+   *           entry by a constant.
+   */
+  Vector<T,dimension>& operator+=(const T rhs);
+  Vector<T,dimension>& operator-=(const T rhs);
+
+  /**
+   * @brief    Scalar multiplication/division of vectors.
+   */
+  Vector<T,dimension>& operator*=(const T rhs);
+  Vector<T,dimension>& operator/=(const T rhs);
+
+ private:
+  std::array<T,dimension> elements_;
+
+};
+
+template <typename T, int dimension>
+Vector<T,dimension>::Vector()
+{
+  for (int i=0; i<dimension; i++)
+    this->elements_[i] = 0;
+}
+
+template <typename T, int dimension>
+Vector<T,dimension>::Vector(const T vector[])
+{
+  for (int i=0; i<dimension; i++)
+    this->elements_[i] = vector[i];
+}
+
+template <typename T, int dimension>
+template <typename U>
+Vector<T,dimension>::Vector(const Vector<U,dimension>& rhs)
+{
+  for (int i=0; i<dimension; i++)
+    this->elements_[i] = T(rhs[i]);
+}
+
+template <typename T, int dimension>
+template <typename U>
+Vector<T,dimension>& Vector<T,dimension>::operator=
+(const Vector<U,dimension>& rhs)
+{
+  for (int i=0; i<dimension; i++)
+    this->elements_[i] = T(rhs[i]);
+  return *this;
+}
+
+template <typename T, int dimension>
+T& Vector<T,dimension>::operator[](int index)
+{
+  return this->elements_[index];
+}
+
+template <typename T, int dimension>
+T Vector<T,dimension>::operator[](int index) const
+{
+  return this->elements_[index];
```

Comment:
No need for `this->`. Member variables are distinguished by the trailing underscore.